tests/Doctrine/Tests/Annotations/Type/ArrayTypeTest.php

```diff
@@ -0,0 +1,222 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\Annotations\Type;
+
+use ArrayObject;
+use Countable;
+use Doctrine\Annotations\Type\ArrayType;
+use Doctrine\Annotations\Type\BooleanType;
+use Doctrine\Annotations\Type\IntegerType;
+use Doctrine\Annotations\Type\IntersectionType;
+use Doctrine\Annotations\Type\MixedType;
+use Doctrine\Annotations\Type\ObjectType;
+use Doctrine\Annotations\Type\StringType;
+use Doctrine\Annotations\Type\UnionType;
+use IteratorAggregate;
+use PHPUnit\Framework\TestCase;
+use stdClass;
+use Traversable;
+
+final class ArrayTypeTest extends TestCase
+{
+    /**
+     * @dataProvider descriptionProvider()
+     */
+    public function testDescription(ArrayType $type, string $description) : void
+    {
+        self::assertSame($description, $type->describe());
+    }
+
+    /**
+     * @param mixed[] $value
+     *
+     * @dataProvider validValuesProvider()
+     */
+    public function testValidValues(ArrayType $type, array $value) : void
+    {
+        self::assertTrue($type->validate($value));
+    }
+
+    /**
+     * @param mixed $value
+     *
+     * @dataProvider invalidValuesProvider()
+     */
+    public function testInvalidValues(ArrayType $type, $value) : void
+    {
+        self::assertFalse($type->validate($value));
+    }
+
+    public function testGetKeyType() : void
+    {
+        $type = new ArrayType(new IntegerType(), new StringType());
+
+        self::assertInstanceOf(IntegerType::class, $type->getKeyType());
+    }
+
+    public function testGetValueType() : void
+    {
+        $type = new ArrayType(new IntegerType(), new StringType());
+
+        self::assertInstanceOf(StringType::class, $type->getValueType());
+    }
+
+    /**
+     * @return (ArrayType|string)[]
+     */
+    public function descriptionProvider() : iterable
```

Comment:
There should be single data provider which includes all: definitions, descriptions, valid values, invalid values. Splitting it makes it very difficult to keep in sync and make sure all cases have correlated tests

Reply:
I don't like the idea of massive multi-purpose data providers, that forces you to focus on things you really are not testing when using such data provider.
Feel free to follow-up with your PR.